ledger/blockHeaderCache.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-deadlock"
+)
+
+const cacheSize = 512
+
+// blockHeaderCache is a wrapper for all block header cache mechanisms used within the Ledger.
+type blockHeaderCache struct {
+	lruCache          heapLRUCache
+	latestHeaderCache latestBlockHeaderCache
+}
+
+type latestBlockHeaderCache struct {
+	blockHeaders [cacheSize]bookkeeping.BlockHeader
+	mutex        deadlock.RWMutex
+}
+
+func (c *blockHeaderCache) initialize() {
+	c.lruCache.maxEntries = 10
+}
+
+func (c *blockHeaderCache) get(round basics.Round) (blockHeader bookkeeping.BlockHeader, exists bool) {
+	// check latestHeaderCache first
+	blockHeader, exists = c.latestHeaderCache.get(round)
+	if exists {
+		return
+	}
+
+	// if not found in latestHeaderCache, check LRUCache
+	value, exists := c.lruCache.Get(round)
+	if exists {
+		blockHeader = value.(bookkeeping.BlockHeader)
+	}
+
+	return
+}
+
+func (c *blockHeaderCache) put(blockHeader bookkeeping.BlockHeader) {
+	c.latestHeaderCache.put(blockHeader)
+	c.lruCache.Put(blockHeader.Round, blockHeader)
+}
+
+func (c *latestBlockHeaderCache) get(round basics.Round) (blockHeader bookkeeping.BlockHeader, exists bool) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	/** round 512 will be stored in idx 511, so the entire required interval for
+	generating the stateproof message will be sequential in memory.
+	Might improve performance in terms of CPU caching.
+	*/
+	idx := basics.SubSaturate(uint64(round), 1) % cacheSize
```

Comment:
I think that the `basics.SubSaturate` here ( and above ) isn't required. I agree that sequential ordering is good, but it doesn't really matter if you shift the array by one or not.

Reply:
round -1 might overflow and I don't really like it. that's why I suggested using `basics.SubSaturate`. @tsachiherman do you have a suggestion of how to keep the block sequentially in memory and not use the -1?

Reply:
yes, just use `idx := round % cacheSize`. It would still keep the sequential ordering, but would not "rotate" the array. It doesn't really give you any advantage to shift it as long as you're doing it symmetrically.

Reply:
In order to construct stateproof message for interval 512 (for example) we need to gather all the headers for rounds : [ 257, 258 ... 512]. If we use `idx := round % cacheSize` we end up with the last round (512) placed in index 0 in the cache. I would like the order of the cache to reflect the order of fetching the message

Reply:
Ah - I see. I think that when you'll get to the point of actually getting these headers, you'll see that it won't help you.
i.e. you can't return a slice of [257:] since the backing array might get changed. Instead, you'll need to create a copy of the backing array... and if you're creating a copy of the backing array, the index plays a smaller part.

Also, you could calculate the index in a more "branchless" way:
```
idx := (round + cacheSize - 1) % cacheSize
```

either way, I'd encourage you to implement all the methods you would need on this cache as well as to document the rationale for non-trivial indexing choices.